File: src/systems/logger.zig
const std = @import("std");
const elt = @import("../common/element.zig");
const val = @import("../common/values/value.zig");
const date = @import("../utils/date.zig");
const ApicaSystem = @import("apica.zig").ApicaSystem;

pub const LoggerSystem = struct {
    apica: *ApicaSystem,
    actual_file: ?std.fs.File,
    actual_date: [10]u8,

    pub fn init(apica: *ApicaSystem, activate: bool) LoggerSystem {
        var buffer: [10]u8 = [10]u8{ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
        if (activate) {
            date.get_actual_date(&buffer);
        }

        if (buffer[0] != 0) {
            var filepath: [15]u8 = undefined;
            _ = std.fmt.bufPrint(&filepath, "logs/{s}", .{buffer}) catch {};
            std.fs.cwd().makeDir(&filepath) catch {};
        }

        return LoggerSystem{ .apica = apica, .actual_file = null, .actual_date = buffer };
    }

    pub fn destroy(self: *LoggerSystem) void {
        if (self.actual_file) |file| {
            file.close();
        }
    }

    pub fn create_file_for(self: *LoggerSystem, app_name: []const u8) void {
        if (self.actual_date[0] == '\x00') {
            return;
        }

        if (self.actual_file) |file| {
            file.close();
            self.actual_file = null;
        }

        const filepath: []u8 = std.fmt.allocPrint(self.apica.get_allocator(), "logs/{s}/{s}.log", .{ self.actual_date, app_name }) catch {
            return;
        };

        self.actual_file = std.fs.cwd().createFile(filepath, .{}) catch null;
    }

    pub fn __logn_error(self: *const LoggerSystem, err_parts: []const []const u8) void {
        if (self.actual_file) |file| {
            file.writeAll("\x1b[31mAPC_ERR: ") catch {};
            for (err_parts) |part| {
                file.writeAll(part) catch {};
            }
            file.writeAll("\x1b[0m\n") catch {};
        }
    }

    fn log_parameters(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element), start: []const u8, end: []const u8) void {
        if (self.actual_file) |file| {
            file.writeAll(start) catch {};

            for (parameters.items) |param| {
                const converted = param.convert(val.ValueKind.String).get_value().String;

                if (converted.get_value()) |string| {
                    var length: usize = 0;
                    while (length < string.len and string[length] != 0) : (length += 1) {}

                    file.writeAll(string[0..length]) catch {};
                } else {
                    file.writeAll("null") catch {};
                }
            }

            file.writeAll(end) catch {};
        }
    }

    pub fn log_info(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[37mINF: ", "\x1b[0m");
    }

    pub fn logn_info(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[37mINF: ", "\x1b[0m\n");
    }

    pub fn log_success(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[32mSUC: ", "\x1b[0m");
    }

    pub fn logn_success(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[32mSUC: ", "\x1b[0m\n");
    }

    pub fn log_warning(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[33mWRN: ", "\x1b[0m");
    }

    pub fn logn_warning(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[33mWRN: ", "\x1b[0m\n");
    }

    pub fn log_error(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[31mERR: ", "\x1b[0m");
    }

    pub fn logn_error(self: *const LoggerSystem, parameters: std.ArrayList(elt.Element)) void {
        self.log_parameters(parameters, "\x1b[31mERR: ", "\x1b[0m\n");
    }
};

File: src/systems/logger.rs
use std::fs::{create_dir_all, File};
use std::io::Write;
use apica_common::bytecodes::ApicaTypeBytecode;
use apica_common::element::Element;
use apica_common::values::value::Value;
use chrono::Local;

pub struct LoggerSystem {
    actual_file: Option<File>,
    actual_date: Option<String>
}

impl LoggerSystem {
    pub fn init(activate: bool) -> LoggerSystem {
        let date = if activate {
            let now = Local::now();
            Some(now.format("%Y-%m-%d").to_string())
        } else { None };

        if let Some(actual_date) = &date {
            let dirpath = format!("logs/{actual_date}");
            let _ = create_dir_all(dirpath).is_err();
        }

        LoggerSystem { actual_file: None, actual_date: date }
    }

    pub fn create_file_for(&mut self, app_name: &str) {
        if let Some(actual_date) = &self.actual_date {
            let filepath = format!("logs/{}/{}.log", actual_date, app_name);
            self.actual_file = if let Ok(file) = File::create(filepath) { Some(file) } else { None };
        }
    }

    pub fn system_logn_success(&mut self, message: String) {
        if let Some(file) = &mut self.actual_file {
            let _ = writeln!(file, "\x1b[32mAPC_SUC: {message}\x1b[0m");
        }
    }

    pub fn system_logn_error(&mut self, message: String) {
        if let Some(file) = &mut self.actual_file {
            let _ = writeln!(file, "\x1b[31mAPC_ERR: {message}\x1b[0m");
        }
    }

    pub fn log_info(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[37mINF: ", "\x1b[0m");
    }

    pub fn logn_info(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[37mINF: ", "\x1b[0m\n");
    }

    pub fn log_success(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[32mSUC: ", "\x1b[0m");
    }

    pub fn logn_success(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[32mSUC: ", "\x1b[0m\n");
    }

    pub fn log_warning(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[33mWRN: ", "\x1b[0m");
    }

    pub fn logn_warning(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[33mWRN: ", "\x1b[0m\n");
    }

    pub fn log_error(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[31mERR: ", "\x1b[0m");
    }

    pub fn logn_error(&mut self, parameters: Vec<Element>) {
        self.log_parameters(parameters, "\x1b[31mERR: ", "\x1b[0m\n");
    }

    fn log_parameters(&mut self, parameters: Vec<Element>, start: &str, end: &str) {
        if let Some(file) = &mut self.actual_file {
            let _ = write!(file, "{start}");
            for param in &parameters {
                if let Value::String(parameter) = param.get_value() {
                    let _ = write!(file, "{}", parameter.get_value().as_ref().unwrap_or(&"null".to_string()));
                } else {
                    let converted = param.convert(ApicaTypeBytecode::String);
                    let parameter = match converted.get_value() {
                        Value::String(s) => s,
                        _ => unreachable!(),
                    };

                    let _ = write!(file, "{}", parameter.get_value().as_ref().unwrap_or(&"null".to_string()));
                }
            }

            let _ = write!(file, "{end}");
        }
    }
}